src/cargo/ops/cargo_compile.rs

```diff
@@ -165,33 +165,66 @@ pub fn compile_pkg<'a>(package: &Package,
         (packages, resolved_with_overrides, registry.move_sources())
     };
 
-    let pkgid = match spec {
-        Some(spec) => try!(resolve_with_overrides.query(spec)),
-        None => package.package_id(),
+    let mut invalid_spec = vec![];
+    let pkgids = if spec.len() > 0 {
+        spec.iter().map(|&p| {
+            let id = resolve_with_overrides.query(p);
+            if id.is_err() {
+                invalid_spec.push(p);
+            }
+            id
+        }).filter(|id|
+            id.is_ok()
+        ).map(|id| id.unwrap()).collect::<Vec<&PackageId>>()
```

Comment:
This may be better expressed as:

```rust
spec.iter().filter_map(|p| {
    match resolved_with_overrides.query(p) {
        Ok(p) => Some(p),
        Err(..) => { invalid_spec.push(p); None }
    }
}).collect::<Vec<_>>();
```

I'm also a little hesitant about throwing away the error information, I think invalid specifications will want to retain the error information about *why* they were invalid